### PR TITLE
Review

### PR DESCRIPTION
**Pull request by MaxleyCv**
This ranch includes at least some sort of project structure

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,23 @@
...
+```
+1  
+1 1  
+1 1 2  
+1 3 1 7  
+```
+### Output example:
+```
+11
+```
```
Comment (PASSTor) [inline README.md:23]: Please, include the algorithm description in your README

File: README.md [hunk 2]
```diff
@@ -1,2 +1,23 @@
-# BFS-IoT
+# Breadth First Search algorithm
 Implementation of variant "carreer" for IoT LPNU Practical #3
+
+## CAREER
+The company has a hierarchy of a pyramidal type.
```
Comment (PASSTor) [inline README.md:5]: You can just copy paste the task here – would be easier to read for reviewers

File: career.out [hunk 3]
```diff
@@ -0,0 +1 @@
+13
```
Comment (PASSTor) [inline career.out:1]: Not a good idea to commit .out files ;)

File: career.in [hunk 4]
```diff
@@ -0,0 +1,4 @@
+3
```
Comment (PASSTor) [inline career.in:1]: You could have used a folder for text files. It's not that important with this number of files, but usually is a good practice

File: graph.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+class Vertex:
+    """
+    Vertex of the graph, has connections that are relative for a particular graph
+    (e.g. one vertex is impossible to use in two different graphs)
+    """
+
+    def __init__(self, val):
+        self.value = val
+        self.connections = []
```
Comment (PASSTor) [inline graph.py:12]: When you design a class – try to answer a question: does it belong here? 
In this case: Does the Vertex has to know about its connections? Usually the answer is "no", but you may still decide differently – just be prepared to argument your choice

File: graph.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+        self.pathway = 0
```
Comment (PASSTor) [inline graph.py:13]: Even more interesting var. For a reviewer it's difficult to understand what this var contains in the Vertex context

File: graph.py [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+
+class Graph:
+    """
+    Class for iplementation of the graphs
+    Has a list of Vertexes, each vertex has value and maximum pathway property
+    """
+
+    def __init__(self):
+        self.vertexes = []
+        self.max_profit = 0
```
Comment (PASSTor) [inline graph.py:24]: Well, if you implement not a generic graph class, but rather a specific one, which knows about max_profit – then name it properly. Something like "CorpHierarchyGraph" will do.

File: graph.py [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+
+    def add_vertex(self, val, connections):
+        """
+        Method adds new vertex to a graph, and adds connections to other vertexes
+        :param val: value of new vertex
+        :param connections: list of connections
+        :return: nothing
+        """
+        new_vertex = Vertex(val)
+        new_vertex.connections = copy.deepcopy(connections)
```
Comment (PASSTor) [inline graph.py:34]: Usually this is not a good idea in a library (and this class is kinda a library class – can be called by other apps potentially, right?). If the user of the code would need a copy – let them decide if they want it when they call your method. 

i.e. in this case it might be better to call `add_vertex(val, copy.deepcopy(connections))` than to do it inside the method.

File: graph.py [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+        for vertex in self.vertexes:
+            vertex.pathway = 0
```
Comment (PASSTor) [inline graph.py:36]: So... With every vertex added you go through all your vertexes again and again and zero the pathways for all of them? :)

File: graph.py [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+        self.vertexes.append(new_vertex)
+
+    def breadth_search(self, indexes, is_first=False):
```
Comment (PASSTor) [inline graph.py:39]: It's breadth *first* search. I know it's not important, but you never know who you will deal with in the next lab :)

File: graph.py [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+        >>> G.add_vertex(2, [0, 1])
+        >>> G.add_vertex(3, [2])
+        >>> G.breadth_search([3], is_first=True)
+        >>> G.max_profit
+        9
+        >>> G.add_vertex(4, [])
+        >>> G.vertexes[3].pathway
+        0
+        """
+        if is_first:
```
Comment (PASSTor) [inline graph.py:59]: Yeah... not a good idea in a non-sport code. How do you think it can be done better?

File: graph.py [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+            self.vertexes[indexes[0]].pathway = self.vertexes[indexes[0]].value
+
+        next_indexes = []
+        for index in indexes:
+            current = copy.deepcopy(self.vertexes[index])
```
Comment (PASSTor) [inline graph.py:64]: And you deepcopy here why exactly?

Comment (PASSTor) [reply graph.py:64]: This is not a very good habit. If you're deepcopying because you don't want to break internal state of the object – maybe there's a better way to design your classes to avoid deepcopy. 

In general, deepcopy is slow and may not be possible in some cases, if the object is not serialisable. I would not recommend to use it too often

File: main.py [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+            if len(new_level) == length:
+                hierarchy.add_vertex(new_level[new_index], [])
+            else:
+                hierarchy.add_vertex(new_level[new_index], generate_indexes(len(new_level), new_index, prev_length))
+            
+    return hierarchy
+
+
+if __name__ == "__main__":
+#    doctest.testmod(verbose=True)
```
Comment (PASSTor) [inline main.py:79]: Doctests are good, but they're not a substitute for unit tests. Don't abuse doctests ;) Unit tests need to check for happy path, edge cases, wrong inputs, etc – if you start creating such tests in docstrings – your code would become a monster

File: main.py [hunk 14]
```diff
@@ -0,0 +1,81 @@
+import doctest
+from graph import *
+
+INPUT_FILE = open("career.in", "r")
```
Comment (PASSTor) [inline main.py:4]: Yeah... Not  a great decision. Who closes this file? And more interesting – when? :)

Bonus question: Do you know why most of the python file examples use `with`?

File: main.py [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+    13
+    >>> G = Graph()
+    >>> G.add_vertex(4, [])
+    >>> G.add_vertex(1, [])
+    >>> G.add_vertex(2, [0, 1])
+    >>> get_maximum_profit(G)
+    6
+    """
+    if len(hierarchy.vertexes):
+        hierarchy.breadth_search([len(hierarchy.vertexes) - 1], is_first=True)
```
Comment (PASSTor) [inline main.py:23]: So you call this method only here. And obviously you want to pass just one value – why the hell is your signature only accepts array then?

Bonus question: why not just -1 instead of `len(hierarchy.vertexes) - 1?`

Comment (PASSTor) [reply main.py:23]: Also, why not to return the value you're interested in? Why instead to store a state inside a class?

File: no_graph_implementation.py [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+
+
+def get_input():
+    """
+    Function to get list of lists of vertexes
+    :return: list of levels
+    """
+    result = []
+    income = INPUT_FILE.readlines()
+    for i in range(len(income)):
```
Comment (PASSTor) [inline no_graph_implementation.py:14]: Why not just `for s in INPUT_FILE.readlines()` ? Or even a `map` instead of cycle? (I guess all you need is string.strip)

File: no_graph_implementation.py [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+    Main function to find maximum
+    :param items: list of levels in company
+    :return: maximum career growth
+    >>> get_max_career([[1], [2, 3], [1, 9, 2]])
+    13
+    >>> get_max_career([[0], [1, 1], [2, 3, 2], [10, 11, 9, 8]])
+    15
+    """
+    indexes = range(1, len(items))
+    indexes = reversed(indexes)
```
Comment (PASSTor) [inline no_graph_implementation.py:33]: Have you ever used third parameter in range()?

**Review by PASSTor — APPROVED** [review]

File: main.py (hunk 15 above)
Comment (MaxleyCv) [reply main.py:23]: Storing value inside of class is of course a nonsense for a method.
But it was more comfortable to do ion the stage of implementation, because of the ease of reaching this value.
However regarding OOP principles, it would be better to return a value from the method (maximum path length does not describe a part of graph and is also not a constant value)

File: main.py (hunk 14 above)
Comment (MaxleyCv) [reply main.py:4]: I forgot to close the file in the end. Honestly, I have never used ```with closing()``` before, however it is mostly used to close the file whenever the runtime ends. And it is a really useful thing, thank you :)

File: main.py (hunk 13 above)
Comment (MaxleyCv) [reply main.py:79]: Okay, I will do them)

File: graph.py (hunk 5 above)
Comment (MaxleyCv) [reply graph.py:12]: Of course the answer is "no", and these connections can be easily moved to Graph class here)

File: career.out (hunk 3 above)
Comment (MaxleyCv) [reply career.out:1]: Okay, they are trash in the repository)
Now I understand that it is not readable)

Comment (PASSTor) [reply career.out:1]: It's not about being readable. It's about results might be false. Imaging I run your algo on my laptop and then git pull (or git reset or whatever else). Suddenly the .out file is overwritten by the one from remote repository. Which means I might be debugging code for wrong reasons. 

As a general rule of thumb: all files that are derivatives from your code (.exe, .pyc, .out) should not be in the repo, unless you really know why you include them

File: main.py (hunk 15 above)
Comment (PASSTor) [reply main.py:23]: Yes and it's not only OOP. In C++ you can mark arguments as IN or OUT, in other languages you don't. A method is usually expected to change a state of self or return the value. In rare cases it changes a state of third party objects. Nevertheless if you think of usability of your code — returning is a good habit
